src/compiler/checker.ts

```diff
@@ -8498,53 +8524,131 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     const identifier = setEmitFlags(factory.createIdentifier(symbolName), EmitFlags.NoAsciiEscaping);
                     if (typeParameterNodes) setIdentifierTypeArguments(identifier, factory.createNodeArray<TypeNode | TypeParameterDeclaration>(typeParameterNodes));
                     identifier.symbol = symbol;
 
                     return index > 0 ? factory.createPropertyAccessExpression(createExpressionFromSymbolChain(chain, index - 1), identifier) : identifier;
                 }
                 else {
-                    if (firstChar === CharacterCodes.openBracket) {
-                        symbolName = symbolName.substring(1, symbolName.length - 1);
-                        firstChar = symbolName.charCodeAt(0);
+                    let hasComputedProperty = false;
+                    if (symbolNameSource && isComputedPropertyName(symbolNameSource)) {
+                        hasComputedProperty = true;
+                        symbolNameSource = symbolNameSource && isComputedPropertyName(symbolNameSource) ? symbolNameSource.expression : undefined;
+                        symbolName = symbolNameSource && isIdentifier(symbolNameSource) ? declarationNameToString(symbolNameSource) : symbolName.substring(1, symbolName.length - 1);
+                        useSingleQuote = !(context.flags & NodeBuilderFlags.UseDoubleQuotesForStringLiteralType) &&
+                            (!!(context.flags & NodeBuilderFlags.UseSingleQuotesForStringLiteralType) || isSingleOrDoubleQuote(symbolName.charCodeAt(0)));
                     }
                     let expression: Expression | undefined;
-                    if (isSingleOrDoubleQuote(firstChar) && !(symbol.flags & SymbolFlags.EnumMember)) {
-                        expression = factory.createStringLiteral(stripQuotes(symbolName).replace(/\\./g, s => s.substring(1)), firstChar === CharacterCodes.singleQuote);
-                    }
-                    else if (("" + +symbolName) === symbolName) {
-                        expression = factory.createNumericLiteral(+symbolName);
-                    }
+
+                    const nameType = getSymbolLinks(symbol).nameType;
+                    const hasEnumOrSymbolName = nameType && nameType.flags & (TypeFlags.EnumLiteral | TypeFlags.UniqueESSymbol);
+                    if (symbolNameSource) {
+                        // If we have a source node for the name it might be string number or entity name that we can reuse
+                        if (isStringLiteral(symbolNameSource)) {
+                            const stringLiteralName = cloneStringLiteral(symbolNameSource, overrideSourceStringDelimiter, useSingleQuote);
+                            stringLiteralName.symbol = symbol;
+                            expression = stringLiteralName;
+                        }
+                        else if (("" + +symbolName) === symbolName && !hasEnumOrSymbolName) {

```

Comment:
You probably want to use `isValidNumberString` to avoid handling a positive or negative `Infinity` literal type (which you can get by overflowing) like a number here.